File: dental-academy-clean/analysis/domain_statistics.js
const fs = require('fs');
const path = require('path');

const analyzeDomains = () => {
  try {
    // Читаем файл с вопросами
    const questionsData = JSON.parse(fs.readFileSync(path.join(__dirname, '../scripts/160_2.json'), 'utf8'));
    
    // Статистика по доменам
    const domainStats = {};
    const domainIRT = {};
    const domainDifficulty = {};
    const domainCategories = {};
    
    // Анализируем каждый вопрос
    questionsData.forEach(question => {
      const domain = question.domain;
      
      // Подсчет вопросов по доменам
      domainStats[domain] = (domainStats[domain] || 0) + 1;
      
      // IRT параметры по доменам
      if (!domainIRT[domain]) {
        domainIRT[domain] = {
          difficulty: [],
          discrimination: [],
          guessing: []
        };
      }
      
      if (question.irt_params) {
        domainIRT[domain].difficulty.push(question.irt_params.difficulty);
        domainIRT[domain].discrimination.push(question.irt_params.discrimination);
        domainIRT[domain].guessing.push(question.irt_params.guessing);
      }
      
      // Уровни сложности по доменам
      if (!domainDifficulty[domain]) {
        domainDifficulty[domain] = { 1: 0, 2: 0, 3: 0 };
      }
      domainDifficulty[domain][question.difficulty_level]++;
      
      // Категории по доменам
      if (!domainCategories[domain]) {
        domainCategories[domain] = new Set();
      }
      domainCategories[domain].add(question.category);
    });
    
    // Вычисляем средние IRT параметры
    const averageIRT = {};
    Object.keys(domainIRT).forEach(domain => {
      const difficulty = domainIRT[domain].difficulty;
      const discrimination = domainIRT[domain].discrimination;
      const guessing = domainIRT[domain].guessing;
      
      averageIRT[domain] = {
        avg_difficulty: difficulty.length > 0 ? (difficulty.reduce((a, b) => a + b, 0) / difficulty.length).toFixed(3) : 0,
        avg_discrimination: discrimination.length > 0 ? (discrimination.reduce((a, b) => a + b, 0) / discrimination.length).toFixed(3) : 0,
        avg_guessing: guessing.length > 0 ? (guessing.reduce((a, b) => a + b, 0) / guessing.length).toFixed(3) : 0,
        min_difficulty: Math.min(...difficulty),
        max_difficulty: Math.max(...difficulty),
        min_discrimination: Math.min(...discrimination),
        max_discrimination: Math.max(...discrimination)
      };
    });
    
    // Преобразуем Set в массивы для JSON
    const domainCategoriesArray = {};
    Object.keys(domainCategories).forEach(domain => {
      domainCategoriesArray[domain] = Array.from(domainCategories[domain]);
    });
    
    // Анализ проблем
    const problems = {
      insufficient_domains: [],
      excessive_domains: [],
      invalid_irt: [],
      duplicate_domains: []
    };
    
    // Домены с недостаточным количеством вопросов
    Object.keys(domainStats).forEach(domain => {
      if (domainStats[domain] < 10) {
        problems.insufficient_domains.push({
          domain: domain,
          count: domainStats[domain],
          needed: 10 - domainStats[domain]
        });
      }
    });
    
    // Домены с избыточным количеством вопросов
    Object.keys(domainStats).forEach(domain => {
      if (domainStats[domain] > 30) {
        problems.excessive_domains.push({
          domain: domain,
          count: domainStats[domain],
          excess: domainStats[domain] - 30
        });
      }
    });
    
    // Проверка IRT параметров
    questionsData.forEach(question => {
      if (question.irt_params) {
        if (question.irt_params.difficulty > 2 || question.irt_params.discrimination < 1) {
          problems.invalid_irt.push({
            id: question.id,
            domain: question.domain,
            difficulty: question.irt_params.difficulty,
            discrimination: question.irt_params.discrimination,
            issue: question.irt_params.difficulty > 2 ? 'difficulty_too_high' : 'discrimination_too_low'
          });
        }
      }
    });
    
    // Поиск дублирующихся доменов
    const potentialDuplicates = {
      'PHARMA': 'FARMACOLOGIE',
      'ETHIEK': 'PROFESSIONAL'
    };
    
    Object.keys(potentialDuplicates).forEach(domain1 => {
      const domain2 = potentialDuplicates[domain1];
      if (domainStats[domain1] && domainStats[domain2]) {
        problems.duplicate_domains.push({
          domain1: domain1,
          domain2: domain2,
          count1: domainStats[domain1],
          count2: domainStats[domain2],
          total: domainStats[domain1] + domainStats[domain2]
        });
      }
    });
    
    // Общая статистика
    const totalQuestions = questionsData.length;
    const totalDomains = Object.keys(domainStats).length;
    const averageQuestionsPerDomain = (totalQuestions / totalDomains).toFixed(2);
    
    const report = {
      summary: {
        total_questions: totalQuestions,
        total_domains: totalDomains,
        average_questions_per_domain: parseFloat(averageQuestionsPerDomain),
        date_analyzed: new Date().toISOString()
      },
      domain_statistics: domainStats,
      domain_irt_averages: averageIRT,
      domain_difficulty_levels: domainDifficulty,
      domain_categories: domainCategoriesArray,
      problems: problems
    };
    
    // Сохраняем отчет
    fs.writeFileSync(path.join(__dirname, 'domain_report.json'), JSON.stringify(report, null, 2));
    
    console.log('✅ Анализ завершен. Результаты сохранены в analysis/domain_report.json');
    
    return report;
    
  } catch (error) {
    console.error('❌ Ошибка при анализе:', error.message);
    return null;
  }
};

// Запускаем анализ
if (require.main === module) {
  analyzeDomains();
}

module.exports = { analyzeDomains }; 

File: dental-academy-clean/analysis/domain_statistics.py
#!/usr/bin/env python3
import json
import os
from collections import defaultdict
from datetime import datetime

def analyze_domains():
    try:
        # Читаем файл с вопросами
        with open('../scripts/160_2.json', 'r', encoding='utf-8') as f:
            questions_data = json.load(f)
        
        # Статистика по доменам
        domain_stats = defaultdict(int)
        domain_irt = defaultdict(lambda: {'difficulty': [], 'discrimination': [], 'guessing': []})
        domain_difficulty = defaultdict(lambda: {1: 0, 2: 0, 3: 0})
        domain_categories = defaultdict(set)
        
        # Анализируем каждый вопрос
        for question in questions_data:
            domain = question['domain']
            
            # Подсчет вопросов по доменам
            domain_stats[domain] += 1
            
            # IRT параметры по доменам
            if 'irt_parameters' in question:
                irt = question['irt_parameters']
                domain_irt[domain]['difficulty'].append(irt['difficulty'])
                domain_irt[domain]['discrimination'].append(irt['discrimination'])
                domain_irt[domain]['guessing'].append(irt['guessing'])
            
            # Уровни сложности по доменам
            domain_difficulty[domain][question['difficulty_level']] += 1
            
            # Категории по доменам
            domain_categories[domain].add(question['category'])
        
        # Вычисляем средние IRT параметры
        average_irt = {}
        for domain, irt_data in domain_irt.items():
            difficulty = irt_data['difficulty']
            discrimination = irt_data['discrimination']
            guessing = irt_data['guessing']
            
            if difficulty:
                average_irt[domain] = {
                    'avg_difficulty': round(sum(difficulty) / len(difficulty), 3),
                    'avg_discrimination': round(sum(discrimination) / len(discrimination), 3),
                    'avg_guessing': round(sum(guessing) / len(guessing), 3),
                    'min_difficulty': min(difficulty),
                    'max_difficulty': max(difficulty),
                    'min_discrimination': min(discrimination),
                    'max_discrimination': max(discrimination)
                }
        
        # Преобразуем Set в списки для JSON
        domain_categories_list = {domain: list(categories) for domain, categories in domain_categories.items()}
        
        # Анализ проблем
        problems = {
            'insufficient_domains': [],
            'excessive_domains': [],
            'invalid_irt': [],
            'duplicate_domains': []
        }
        
        # Домены с недостаточным количеством вопросов
        for domain, count in domain_stats.items():
            if count < 10:
                problems['insufficient_domains'].append({
                    'domain': domain,
                    'count': count,
                    'needed': 10 - count
                })
        
        # Домены с избыточным количеством вопросов
        for domain, count in domain_stats.items():
            if count > 30:
                problems['excessive_domains'].append({
                    'domain': domain,
                    'count': count,
                    'excess': count - 30
                })
        
        # Проверка IRT параметров
        for question in questions_data:
            if 'irt_parameters' in question:
                irt = question['irt_parameters']
                if irt['difficulty'] > 2 or irt['discrimination'] < 1:
                    problems['invalid_irt'].append({
                        'id': question['id'],
                        'domain': question['domain'],
                        'difficulty': irt['difficulty'],
                        'discrimination': irt['discrimination'],
                        'issue': 'difficulty_too_high' if irt['difficulty'] > 2 else 'discrimination_too_low'
                    })
        
        # Поиск дублирующихся доменов
        potential_duplicates = {
            'PHARMA': 'FARMACOLOGIE',
            'ETHIEK': 'PROFESSIONAL'
        }
        
        for domain1, domain2 in potential_duplicates.items():
            if domain1 in domain_stats and domain2 in domain_stats:
                problems['duplicate_domains'].append({
                    'domain1': domain1,
                    'domain2': domain2,
                    'count1': domain_stats[domain1],
                    'count2': domain_stats[domain2],
                    'total': domain_stats[domain1] + domain_stats[domain2]
                })
        
        # Общая статистика
        total_questions = len(questions_data)
        total_domains = len(domain_stats)
        average_questions_per_domain = round(total_questions / total_domains, 2)
        
        report = {
            'summary': {
                'total_questions': total_questions,
                'total_domains': total_domains,
                'average_questions_per_domain': average_questions_per_domain,
                'date_analyzed': datetime.now().isoformat()
            },
            'domain_statistics': dict(domain_stats),
            'domain_irt_averages': average_irt,
            'domain_difficulty_levels': dict(domain_difficulty),
            'domain_categories': domain_categories_list,
            'problems': problems
        }
        
        # Сохраняем отчет
        with open('domain_report.json', 'w', encoding='utf-8') as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        
        print('✅ Анализ завершен. Результаты сохранены в analysis/domain_report.json')
        
        return report
        
    except Exception as error:
        print(f'❌ Ошибка при анализе: {error}')
        return None

if __name__ == '__main__':
    analyze_domains() 